### PR TITLE
Potentially more efficient way to filter virtual aps

### PR DESCRIPTION
**Pull request by gcleary**
Please take a look at this possible better way to handle filter_virtual_aps()

Rather than waiting till the workspace is full of aps, before starting to filter them (which requires a search across all possible combinations of the beacons in the workspace, we can check each one added (the way we already remove identical duplicates) as it is added.

This is a little more expensive if the number of aps added is less than the number we can hold in the workspace, but is much more efficient for every beacon added there after,

Overall, I think it is worth doing. I think it simplifies the logic of the filtering.

### CODE REVIEW

File: libel/beacons.c [hunk 1]
```diff
@@ -226,74 +226,73 @@ static Sky_status_t filter_by_rssi(Sky_ctx_t *ctx)
     return remove_beacon(ctx, reject);
 }
 
-/*! \brief try to reduce AP by filtering out virtual AP
+/*! \brief try to reduce AP by filtering out one virtual AP that matches beacon just added
```
Comment (tboinske-skyhook) [inline libel/beacons.c:229]: "reduce AP": perhaps "reduce the number of APs"? And why "filtering out"? Why not just "removing"?

File: libel/beacons.c [hunk 2]
```diff
@@ -226,74 +226,73 @@ static Sky_status_t filter_by_rssi(Sky_ctx_t *ctx)
...
+    LOGFMT(ctx, SKY_LOG_LEVEL_DEBUG, "just added beacon at idx %d, ap_len: %d APs of %d beacons", b,
+        (int)ctx->ap_len, (int)ctx->len)
 
     dump_workspace(ctx);
 
-    if (ctx->ap_len <= CONFIG(ctx->cache, max_ap_beacons)) {
-        return false;
-    }
-
     /* look for any AP beacon that is 'similar' to another */
```
Comment (tboinske-skyhook) [inline libel/beacons.c:252]: This comment appears to be out of place. It probably should be combined with the comment at line 258?

File: libel/beacons.c [hunk 3]
```diff
@@ -226,74 +226,73 @@ static Sky_status_t filter_by_rssi(Sky_ctx_t *ctx)
...
  *         When similar, remove beacon with highesr mac address
  *         unless it is in cache, then choose to remove the uncached beacon
  *
  *  @param ctx Skyhook request context
+ *  @param b index of beacon just added
  *
  *  @return true if beacon removed or false otherwise
  */
-static bool filter_virtual_aps(Sky_ctx_t *ctx)
+static bool filter_virtual_aps(Sky_ctx_t *ctx, int b)
```
Comment (tboinske-skyhook) [inline libel/beacons.c:238]: Would `remove_duplicate_ap()` now be a better name?

**Comment by gcleary** [comment]
> I think the general idea is reasonable, although we should verify that it gives the same result as the previous method. Approved, with some commenting and renaming suggestions.

The results are not identical with this approach. The old algorithm search for virtual APs after adding the new AP in rssi order and after the workspace was full. So there could be multiple matching virtual APs the first time the search happened. With the new approach, no virtual matches ever make it into the workspace.
However, for the two approaches, the actual beacons that are included in the request to the server for a given scan, should be the same, if any virtual matching APs are ignored.

**Comment by tboinske-skyhook** [comment]
> However, for the two approaches, the actual beacons that are included in the request to the server for a given scan, should be the same, if any virtual matching APs are ignored.

**Comment by tboinske-skyhook** [comment]
Not sure I follow. But I guess I'll amend my remark by saying instead that we should verify that the results that the new algo gives are at least as "desirable" as the those given by the previous algo.

**Comment by gcleary** [comment]
> Not sure I follow. But I guess I'll amend my remark by saying instead that we should verify that the results that the new algo gives are at least as "desirable" as the those given by the previous algo.

Yes, agreed. I'm making the point that the old algorithm could leave some virtual aps in the request, but the new approach will always remove them.

File: libel/beacons.c (hunk 1 above)
Comment (gcleary) [reply libel/beacons.c]: Yes. Done.

File: libel/beacons.c (hunk 2 above)
Comment (gcleary) [reply libel/beacons.c]: Yes. Done.

**Pull request merged**